test/common/upstream/cluster_manager_impl_test.cc

```diff
@@ -203,6 +206,255 @@ envoy::config::bootstrap::v3::Bootstrap defaultConfig() {
+
+  handle2.reset();
+  handle3.reset();
+  handle1.reset();
+  handle4.reset();
+
+  cluster_manager_->addOrUpdateCluster(defaultStaticCluster("cluster_foo1"), "version1");
+  cluster_manager_->addOrUpdateCluster(defaultStaticCluster("cluster_foo2"), "version1");
+  cluster_manager_->addOrUpdateCluster(defaultStaticCluster("cluster_foo3"), "version1");
+  cluster_manager_->addOrUpdateCluster(defaultStaticCluster("cluster_foo4"), "version1");
+}
+
+// Checks that dropping discovery handles will result in callbacks not being invoked.
+TEST_F(ODCDTest, TestHandles) {
+  auto cb1 = createCallback(ClusterDiscoveryStatus::Available);
+  auto cb2 = std::make_unique<ClusterDiscoveryCallback>(
+      [](ClusterDiscoveryStatus) { ADD_FAILURE() << "The callback 2 should not be called"; });
+  auto cb3 = std::make_unique<ClusterDiscoveryCallback>(
+      [](ClusterDiscoveryStatus) { ADD_FAILURE() << "The callback 3 should not be called"; });
+  auto cb4 = createCallback(ClusterDiscoveryStatus::Available);
+  EXPECT_CALL(*odcds_, updateOnDemand("cluster_foo"));
+  auto handle1 =
+      odcds_handle_->requestOnDemandClusterDiscovery("cluster_foo", std::move(cb1), timeout_);
+  auto handle2 =
+      odcds_handle_->requestOnDemandClusterDiscovery("cluster_foo", std::move(cb2), timeout_);
+  auto handle3 =
+      odcds_handle_->requestOnDemandClusterDiscovery("cluster_foo", std::move(cb3), timeout_);
+  auto handle4 =
+      odcds_handle_->requestOnDemandClusterDiscovery("cluster_foo", std::move(cb4), timeout_);
+
+  // handle1 and handle4 are left intact, so their respective callbacks will be invoked.
+  handle2.reset();
+  handle3.reset();
+
+  cluster_manager_->addOrUpdateCluster(defaultStaticCluster("cluster_foo"), "version1");
+  EXPECT_EQ(callback_call_count_, 2);
+}
+
+// Check if callback is invoked when trying to discover a cluster we already know about. It should
+// not call into ODCDS in such case.
+TEST_F(ODCDTest, TestCallbackWithExistingCluster) {
+  auto cb = createCallback(ClusterDiscoveryStatus::Available);
+  cluster_manager_->addOrUpdateCluster(defaultStaticCluster("cluster_foo"), "version1");
+  EXPECT_CALL(*odcds_, updateOnDemand("cluster_foo")).Times(0);
+  auto handle =
+      odcds_handle_->requestOnDemandClusterDiscovery("cluster_foo", std::move(cb), timeout_);
+  EXPECT_EQ(callback_call_count_, 1);
+}
+
+// Checks that the cluster manager detects that a thread has requested a cluster that some other
+// thread already did earlier, so it does not start another discovery process.
+TEST_F(ODCDTest, TestMainThreadDiscoveryInProgressDetection) {
+  EXPECT_CALL(*odcds_, updateOnDemand("cluster_foo"));
+  auto cb1 = createCallback();
+  auto cb2 = createCallback();
+  auto handle1 =
+      odcds_handle_->requestOnDemandClusterDiscovery("cluster_foo", std::move(cb1), timeout_);
+  auto cdm = cluster_manager_->createAndSwapClusterDiscoveryManager("another_fake_thread");
```

Comment:
It is mostly an emulation thing. Emulating threads in these unit tests is icky. Each worker thread has its own cluster discovery manager (CDM). Here I emulate having a second thread by creating another CDM with a different name. I issued a discovery request in one "thread" (CDM, really) for a certain cluster and then did the same in another "thread". This is to test the case where only main thread can decide whether a discovery for a cluster was already requested and it's ongoing.

I probably don't need the old CDM to be around, but there are things that happen implicitly here when objects go out of scope and their destructors get invoked - callbacks being removed and so on. I erred here on the side of caution and kept the first CDM alive just as it would be with real threads.